### PR TITLE
Add spell check tools to eng/common

### PR DESCRIPTION
**Pull request by danieljurek**
See example run in Embedded C repo -- https://dev.azure.com/azure-sdk/public/_build/results?buildId=739511&view=logs&j=a129effc-2dd1-54d1-fb5a-ad7bdc0e851d&t=7b4713e3-e110-5049-e231-cc8a1e664c68

Logs: 
image

Pipeline warnings: 
image

check-spelling.yml 
 * Runs only on PRs 
 * Fetches the "base" ref (usually the `master` branch) of the PR.. this takes some work as the git config in DevOps is not set to match branch names the way a typical `git clone` would 
 * Runs `Test-Spelling.ps1` 
 * Needs to be wired into a build job

Test-Spelling.ps1 is designed to run on a dev box as well as in DevOps. It will advise if there are errors. 

I would have preferred to put `cspell.json` in the `eng/` folder in each repo. However the [configuration in the vcsode plugin](https://github.com/streetsidesoftware/vscode-spell-checker/blob/master/packages/client/README.md#customization) searches only in the root of the repo, the `.vscode` folder, or specific preferences set in `.vscode/settings.json` (or at a user level). If we want to keep the root of the repo less cluttered then it makes sense to put `cspell.json` into `.vscode` (see [demo PR](https://github.com/Azure/azure-sdk-for-c/pull/1610/files#diff-d9f1a54e1f1a4c8324f4756f08a6af27671a3ae9312190d4a5fe981cc1722824)). 

Of note: there are [notes in the vscode extension repo ](https://github.com/streetsidesoftware/vscode-spell-checker/blob/master/packages/client/docs/settings.md) about possibly using a `cspell.configLocation` setting but that functionality does not yet exist.

### CODE REVIEW

File: eng/common/scripts/Test-Spelling.ps1 [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+$initialDirectory = Get-Location
+$exitCode = 0
+try { 
+    Set-Location "$PSScriptRoot/../../.."
+
+    # Lists names of files that were in some way changed between the 
+    # current ref and $TargetRef. Excludes files that were deleted to
+    # prevent errors in Resolve-Path
+    Write-Host "git diff --diff-filter=d --name-only $TargetRef"
```
Comment (benbp) [inline eng/common/scripts/Test-Spelling.ps1:28]: It's probably worth adding a `git` command check similar to the `npx` check since this is going to be running from powershell sessions.

**Review by benbp — APPROVED** [review]

File: eng/common/pipelines/templates/steps/check-spelling.yml [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+# CspellConfigPath - Path to cspell.json config location
+
+parameters:
+  ContinueOnError: true
+  TargetRef: $(System.PullRequest.TargetBranch)
+  CspellConfigPath: ./.vscode/cspell.json
+
+steps:
+  - ${{ if eq(variables['Build.Reason'], 'PullRequest') }}:
+    # Devops uses a git config that changes ref names.
```
Comment (weshaggard) [inline eng/common/pipelines/templates/steps/check-spelling.yml:16]: Is there any reason the approach that we are using in https://github.com/Azure/azure-sdk-tools/pull/1415 (and https://github.com/Azure/azure-sdk-tools/blob/master/eng/common/scripts/get-markdown-files-from-changed-files.ps1#L5) will not work here? I'd like to use similar strategy across use cases so we aren't doing this differently. It is starting to seem like we need a common way to get a list of the changed files.

File: eng/common/scripts/Test-Spelling.ps1 [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    $changedFilesCount = ($changedFiles | Measure-Object).Count
+    Write-Host "Git Detected $changedFilesCount changed file(s). Files checked by cspell may exclude files according to cspell.json"
+
+    if ($changedFilesCount -eq 0) {
+        Write-Host "No changes detected"
+        # The finally block still runs after calling exit here
+        exit $exitCode
+    }
+
+    $changedFilesString = $changedFiles | Join-String -Separator ' '
```
Comment (weshaggard) [inline eng/common/scripts/Test-Spelling.ps1:45]:
```suggestion
    $changedFilesString = $changedFiles -join ' '
```
While amounts to the same there is a join operator (https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_join?view=powershell-7.1) which is a little more efficient since it doesn't require the use of the object pipeline.

File: eng/common/scripts/Test-Spelling.ps1 [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+if ((Get-Command npx | Measure-Object).Count -eq 0) { 
+    LogError "Could not locate npx. Install NodeJS (includes npm and npx) https://nodejs.org/en/download/"
+    exit 1
+}
+
+$initialDirectory = Get-Location
+$exitCode = 0
+try { 
+    Set-Location "$PSScriptRoot/../../.."
```
Comment (weshaggard) [inline eng/common/scripts/Test-Spelling.ps1:27]: Is there a reason we need to change directories as opposed to just running this script from the correct directory?

File: tools/notification-configuration/common/Services/AzureDevOpsService.cs [hunk 5]
```diff
@@ -179,7 +179,9 @@ public async Task<string> GetDescriptorForPrincipal(string userPrincipal)
         {
             var client = await GetClientAsync<GraphHttpClient>();
 
-            logger.LogInformation("GetDescriptorForAlias UserPrincipal = {0}", userPrincipal);
+            // Do not log PII: "alias@microsoft.com" => "***@microsoft.com"
+            var loggableUserPrincipal = $"***{userPrincipal.Substring(userPrincipal.IndexOf("@"))}";
```
Comment (weshaggard) [inline tools/notification-configuration/common/Services/AzureDevOpsService.cs:183]: This changes seems unrelated.

File: eng/common/scripts/Test-Spelling.ps1 [hunk 6]
```diff
@@ -0,0 +1,87 @@
+[CmdletBinding()]
+Param (
+    [Parameter()]
+    [ValidateNotNullOrEmpty()]
+    [string] $TargetRef = 'master', 
+
+    [Parameter()]
+    [ValidateNotNullOrEmpty()]
+    [string] $CspellConfigPath = "./.vscode/cspell.json"
```
Comment (weshaggard) [inline eng/common/scripts/Test-Spelling.ps1:9]: Does anything bad happen here if this file doesn't exist?

File: eng/common/pipelines/templates/steps/check-spelling.yml [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+# and some ref (branch, tag, etc.) or commit hash. Only runs on PRs.
+# ContinueOnError - true: Pipeline warns on spelling error 
+#                   false: Pipeline fails on spelling error
+# TargetRef       - Target ref (e.g. master) to compare to create file change
+#                   list
+# CspellConfigPath - Path to cspell.json config location
+
+parameters:
+  ContinueOnError: true
+  TargetRef: $(System.PullRequest.TargetBranch)
```
Comment (weshaggard) [inline eng/common/pipelines/templates/steps/check-spelling.yml:11]: Do we expect someone to change this as part of the template yml? Perhaps we should just use the variable and guard against running this step outside of a pull request.

File: eng/common/scripts/Test-Spelling.ps1 [hunk 8]
```diff
@@ -0,0 +1,87 @@
+[CmdletBinding()]
+Param (
+    [Parameter()]
+    [ValidateNotNullOrEmpty()]
+    [string] $TargetRef = 'master',
```
Comment (weshaggard) [inline eng/common/scripts/Test-Spelling.ps1:5]: I'd prefer we default to something like https://github.com/Azure/azure-sdk-tools/blob/master/eng/common/scripts/get-markdown-files-from-changed-files.ps1#L5 instead of master, especially given our upcoming plans to change the default branch to main.

File: eng/common/scripts/Test-Spelling.ps1 (hunk 4 above)
Comment (danieljurek) [reply eng/common/scripts/Test-Spelling.ps1:27]: The only reason would be if you were running from a different directory than the root of the repo. However this isn't a particularly interesting scenario.

File: eng/common/scripts/Test-Spelling.ps1 (hunk 6 above)
Comment (danieljurek) [reply eng/common/scripts/Test-Spelling.ps1:9]: cspell will exit with an error if there is no config. To that end we could validate presence of the file and exit early if it's not there. 

```
> cspell --config doesnotexist.json **/*
Configuration Error: Failed to resolve file: "doesnotexist.json"
CSpell: Files checked: 0, Issues found: 0 in 0 files
```

File: eng/common/pipelines/templates/steps/check-spelling.yml (hunk 7 above)
Comment (danieljurek) [reply eng/common/pipelines/templates/steps/check-spelling.yml:11]: The default is sufficient. This is added for configurability to prevent someone from having to touch eng/common later.

File: eng/common/scripts/Test-Spelling.ps1 (hunk 8 above)
Comment (danieljurek) [reply eng/common/scripts/Test-Spelling.ps1:5]: GitHub doesn't let me use more thumbs up here... 👍 👍 👍

File: eng/common/pipelines/templates/steps/check-spelling.yml (hunk 2 above)
Comment (danieljurek) [reply eng/common/pipelines/templates/steps/check-spelling.yml:16]: Indeed it does work: https://dev.azure.com/azure-sdk/public/_build/results?buildId=741805&view=logs&j=a129effc-2dd1-54d1-fb5a-ad7bdc0e851d&t=617ceecd-bf48-5ce3-83e2-7468b1632a13&l=13 

My concern here is that this approach couples the pipeline logic to how DevOps configures the repo today. My proposal is not as coupled. In the interest of consistency and simplicity I'll use the approach we're already using. If we want to use what I proposed here we can undertake that in a separate effort.

File: eng/common/scripts/Test-Spelling.ps1 (hunk 8 above)
Comment (danieljurek) [reply eng/common/scripts/Test-Spelling.ps1:5]: Because this script is written to be run on a dev box as well I wanted defaults that would work today. In this case I think removing the default is the best course of action so you don't end up with a default value of `origin/` when running on a dev box. 

The yaml template invoking the script has flexibility to put the context-appropriate variables in place.

File: tools/notification-configuration/common/Services/AzureDevOpsService.cs (hunk 5 above)
Comment (danieljurek) [reply tools/notification-configuration/common/Services/AzureDevOpsService.cs]: Was a strange rebase artifact. Reverted. It'll go away in the squash.

File: eng/common/pipelines/templates/steps/check-spelling.yml (hunk 2 above)
Comment (weshaggard) [reply eng/common/pipelines/templates/steps/check-spelling.yml]: I kind of agree with you about the coupling to Devops, in particular the particular variables that Devops sets. We should be able to decouple that in the future by coupling that target branch outside of the script and then passing in the new value. It might be worth filing a follow-up issue to try and refactor that logic out and share it across our different usages.

File: eng/common/scripts/Test-Spelling.ps1 [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+    exit 1
+}
+
+$exitCode = 0
+
+# Lists names of files that were in some way changed between the 
+# current ref and $TargetRef. Excludes files that were deleted to
+# prevent errors in Resolve-Path
+Write-Host "git diff --diff-filter=d --name-only $TargetRef"
+$changedFiles = git diff --diff-filter=d --name-only $TargetRef `
```
Comment (weshaggard) [inline eng/common/scripts/Test-Spelling.ps1:35]: Just to ensure you only get the files you want you should pass HEAD as the other comparison point. Without that it would also potentially pick-up other changes to files in the working directory, etc. 
```suggestion
$changedFiles = git diff --diff-filter=d --name-only $TargetRef HEAD `
```

Comment (danieljurek) [reply eng/common/scripts/Test-Spelling.ps1:35]: 2 parameters: 
* TargetRef -> rename to BaseRef
* Something to hold HEAD so we can insert it in DevOps

File: eng/common/scripts/Test-Spelling.ps1 [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+    exit $exitCode
+}
+
+$changedFilesString = $changedFiles -join ' '
+
+Write-Host "npx cspell --config $CspellConfigPath $changedFilesString"
+$spellingErrors = Invoke-Expression "npx cspell --config $CspellConfigPath $changedFilesString"
+
+if ($spellingErrors) {
+    $exitCode = 1
```
Comment (weshaggard) [inline eng/common/scripts/Test-Spelling.ps1:52]: Lets make the exitcode 0 for these so we don't turn things to red.

File: eng/common/pipelines/templates/steps/check-spelling.yml (hunk 2 above)
Comment (danieljurek) [reply eng/common/pipelines/templates/steps/check-spelling.yml]: https://github.com/Azure/azure-sdk-tools/issues/1437

File: eng/common/scripts/Test-Spelling.ps1 [hunk 11]
```diff
@@ -0,0 +1,91 @@
+[CmdletBinding()]
```
Comment (weshaggard) [inline eng/common/scripts/Test-Spelling.ps1:1]: It might be worth naming this script after what is actually doing. Something like "check-spelling-in-changed-files.ps1" or something along those lines.

File: eng/common/scripts/check-spelling-in-changed-files.ps1 [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+check spelling of files that have been committed and exlcude any new files or
+modified files that are not committed. This is most useful in CI scenarios where
+builds may have modified the state of the repo.
+
+.PARAMETER CspellConfigPath
+Optional location to use for cspell.json path. Default value is 
+`./.vscode/cspell.json`
+
+.EXAMPLE
+./eng/common/scripts/Test-Spelling.ps1 -TargetBranch 'target_branch_name'
```
Comment (weshaggard) [inline eng/common/scripts/check-spelling-in-changed-files.ps1:94]: forgot to update name here.

File: eng/common/scripts/check-spelling-in-changed-files.ps1 [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+    exit 1
+}
+
+if (!(Test-Path $CspellConfigPath)) {
+    LogError "Could not locate config file $CspellConfigPath"
+    exit 1
+}
+
+# Lists names of files that were in some way changed between the 
+# current ref and $TargetBranch. Excludes files that were deleted to
```
Comment (weshaggard) [inline eng/common/scripts/check-spelling-in-changed-files.ps1:33]: comment is a little outdated now.

**Review by weshaggard — APPROVED** [review]
I couple small clean-up comments but otherwise looks good.

File: eng/common/pipelines/templates/steps/check-spelling.yml [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+  CspellConfigPath: ./.vscode/cspell.json
+
+steps:
+  - ${{ if eq(variables['Build.Reason'], 'PullRequest') }}:
+    - task: PowerShell@2
+      displayName: Check spelling (cspell)
+      continueOnError: ${{ parameters.ContinueOnError }}
+      inputs: 
+        targetType: filePath
+        filePath: eng/common/scripts/check-spelling-in-changed-files.ps1
```
Comment (mitchdenny) [inline eng/common/pipelines/templates/steps/check-spelling.yml:22]: nit: [Verb-Noun.ps1] is adiomatic for PowerShell entrypoint scripts.

File: eng/common/scripts/check-spelling-in-changed-files.ps1 [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+
+    [Parameter()]
+    [ValidateNotNullOrEmpty()]
+    [string] $CspellConfigPath = "./.vscode/cspell.json"
+)
+
+$ErrorActionPreference = "Continue"
+. $PSScriptRoot/logging.ps1
+
+if ((Get-Command git | Measure-Object).Count -eq 0) {
```
Comment (mitchdenny) [inline eng/common/scripts/check-spelling-in-changed-files.ps1:17]: Gotta say I love this attention to detail ;)

**Review by mitchdenny — APPROVED** [review]

File: eng/common/scripts/check-spelling-in-changed-files.ps1 [hunk 16]
```diff
@@ -0,0 +1,99 @@
+[CmdletBinding()]
+Param (
+    [Parameter()]
+    [string] $TargetBranch,
+
+    [Parameter()]
+    [string] $SourceBranch,
+
+    [Parameter()]
+    [ValidateNotNullOrEmpty()]
```
Comment (mitchdenny) [inline eng/common/scripts/check-spelling-in-changed-files.ps1:10]: Is ```[ValidateNotNullOrEmpty()]``` helpful here since you set a default? I'd probably avoid setting a default for this particular script since ```./``` assumes you are invoking at the root of the source tree, but if you invoke from within ```eng/common/scripts``` it isn't going to work.

File: eng/common/pipelines/templates/steps/check-spelling.yml (hunk 14 above)
Comment (danieljurek) [reply eng/common/pipelines/templates/steps/check-spelling.yml:22]: @weshaggard initially proposed changing it. I had it called `Test-Spelling.ps1` initially... I hold very lightly the opinion that pouring our scripts into the verb model makes things easier for all end users.

File: eng/common/scripts/check-spelling-in-changed-files.ps1 (hunk 15 above)
Comment (danieljurek) [reply eng/common/scripts/check-spelling-in-changed-files.ps1:17]: @weshaggard deserves credit for the idea of checking for these scripts. @benbp for ensuring that we are consistent in checking for `git`

File: eng/common/scripts/check-spelling-in-changed-files.ps1 (hunk 16 above)
Comment (danieljurek) [reply eng/common/scripts/check-spelling-in-changed-files.ps1:10]: At first I started down the path of coupling the script to its relative location in the repo directory structure (https://github.com/Azure/azure-sdk-tools/pull/1421/commits/b7ddb14f4f7c32880d83aadc17446fbdadce6ca1#diff-331922178311e6d75106a281f4a4e07841c3de3b514f77c5dbc766616ac4fc16L24). But it makes things more complex. The approach we're taking today gives us more flexibility. 

`ValidateNotNullOrEmpty` is only marginally helpful as it would require that some kind of path be fed into the `Test-Path` step a few lines into the script. I don't think its presence takes away from what we're doing here but if it complicates things we can take it out.

File: eng/common/pipelines/templates/steps/check-spelling.yml (hunk 14 above)
Comment (weshaggard) [reply eng/common/pipelines/templates/steps/check-spelling.yml:22]: https://docs.microsoft.com/en-us/powershell/scripting/developer/cmdlet/approved-verbs-for-windows-powershell-commands?view=powershell-7.1 that is the guidance for the verb-noun model but I think makes some sense for cmdlets but  I personally don't think it makes sense for functions or script names. Trying to fit that model is pretty difficult and I think it usually ends up with less understandability for our usages than just naming a script or a function like you would want to call it. So I don't recommend using that guideline for most of our scenarios.
